### ISSUE TITLE
I can't get this to work

### ISSUE THREAD

**Issue by Pauligrinder** [norman-arch/NanoPi-NEO3#1]
I didn't find any other way of contacting you, so I had to write an issue about this:

Anyway, I did exactly according to the instructions for the Rock64 except the dd-steps for this bootloader, and it just doesn't boot. 
At first I thought my Nanopi was faulty, but I managed to get it to boot Ubuntu Core. I also managed to make it boot alarm by keeping the Ubuntu bootloader and formatting the rest of the sd and putting the arch files in there, but then it still stayed on the same 5.4.x kernel even when I updated it on Arch, which caused modules not to load. 

The only thing I can think of is that some offset should be different from the rock64... Could you kindly provide the exact commands you used to make this run?

**Comment by norman-arch** [comment]
if you do have a usb to ttl adapter, you can check the boot info of uboot and then compare with my one, if you see starting kernel ..., then should work.
![IMG_1096](https://user-images.githubusercontent.com/69657857/101238544-ec1e8580-371b-11eb-9d38-dd00444997a3.JPG)
![IMG_1097](https://user-images.githubusercontent.com/69657857/101238554-f80a4780-371b-11eb-8079-2af268a50fdf.PNG)

**Comment by Pauligrinder** [comment]
Thanks for the answer, I'll see if I can get mine hooked up. I have a usb-serial adapter that I guess should do the trick, but I've never used it for anything like this.

One difference I see, is that you have the 2GB model and I have 1GB, maybe that could cause some issue?

I got mine running now using the bootloader from the Ubuntu Core system, but that loads the wrong kernel so I need to do something about it still.

**Comment by stefanvanherwijnen** [comment]
The compiled bootloaders from this repo seem to work fine for me initially.
However, the "Install the U-Boot package" step seems to break things (i.e. the bootloader).

Any clue why?

Edit:
Also `(17/26) upgrading linux-firmware                      [##########################--]  96%` makes everything stop working until I power cycle it. Then after finishing the upgrade it is unreachable after reboot.

**Comment by norman-arch** [comment]
u-boot v2020.07 will cause kernel panic, please update to u-boot v2020.10. With this version, on my neo3, no more kernel panic already more than one month.

**Comment by stefanvanherwijnen** [comment]
I tried a reinstall with the updated u-boot v2020.10 but this time it doesn't boot at all. Ethernet lights are flickering but I can't find it on my network or connect to it :confused: .

**Comment by norman-arch** [comment]
please post error information, otherwise no help to identify root cause.

**Comment by stefanvanherwijnen** [comment]
There don't seem to be any logs at all. I `chrooted` into the partition of the SD card:
```
[root@archlinux log]# journalctl
No journal files were found.
-- No entries --
```
and `/var/log/` seems to be empty. Any suggestions on where to find the errors?

Edit: Unfortunately, I don't have a way to connect the UART.

**Comment by norman-arch** [comment]
please check my previous answer, you have to check u-boot errors via serial console, this is the only way...

**Comment by Pauligrinder** [comment]
> I tried a reinstall with the updated u-boot v2020.10 but this time it doesn't boot at all. Ethernet lights are flickering but I can't find it on my network or connect to it 😕 .

That sounds like what happened to me before.
I've just been happily running mine on the stock Ubuntu Core kernel and haven't had any issues so far (also had to copy the kernel modules), but that's obviously a suboptimal solution... 
I have a usb-uart adapter but haven't gotten around to debugging the situation yet. The main problem being that I don't have a spare microsd card to test on and I don't want to break my working setup, and also I haven't had time to figure out which pins the NanoPi debug interface should connect to (since my adapter has some RS-232 based connector). 

Well, actually, now that I think of it, I do have several microsd's in cameras I could borrow, so I guess it's just an excuse...

It's quite low on my priority list since it works as is, but I will try to get to it at some point.

**Comment by stefanvanherwijnen** [comment]
I've also resorted back to UbuntuCore but I have ordered an UART adapter and will report back with the errors.

Most likely UbuntuCore uses an older uboot so it will have the same problem in the future if it doesn't get fixed upstream.

**Comment by Pauligrinder** [comment]
> I've also resorted back to UbuntuCore but I have ordered an UART adapter and will report back with the errors.
> 
> Most likely UbuntuCore uses an older uboot so it will have the same problem in the future if it doesn't get fixed upstream.

I'm only using the bootloader from UbuntuCore and since I'm not using its' package manager / other services, it never updates the bootloader, so no risk of breakage there 😀  

How I did it, is that I flashed the UbuntuCore image onto a memory card, then wiped the root partition and extracted the Arch Linux root image on there. Then it just worked, apart from anything that relies on a kernel module, which obviously aren't found since Arch comes with a different kernel. So I copied those as well from the UC image.

But like I said before, running an old kernel and bootloader isn't exactly the optimal solution, so I will have to look into it at some point.

I just realized that the UART adapter I mentioned I have before isn't actually an UART adapter, but an AVR JTAG adapter, so I don't know if that'll work in this case after all... 
But, I do have several Raspberry Pi's which have UART directly (via some GPIO pins), so hopefully I'll be able to use one of them for this case (just need to scavenge some old motherboard cables for the pin header connectors... )

**Comment by stefanvanherwijnen** [comment]
Edit: It seems to work after all. I cloned the repo instead of downloading the files manually and that seemed to have solved the boot problems :unamused: .
For future reference, the baud rate needs to be set to 1500000.

I'll complete the install later to see if it keeps working.

**Comment by Pauligrinder** [comment]
> Edit: It seems to work after all. I cloned the repo instead of downloading the files manually and that seemed to have solved the boot problems 😒 .
> For future reference, the baud rate needs to be set to 1500000.
> 
> I'll complete the install later to see if it keeps working.

Cool, I'll try this too at some point. I'm not 100% sure whether I downloaded the files directly or cloned, I may have downloaded them directly too. Since the data is being written using dd, I could've written some 404 error html into the boot area of the sd for all I know...
@norman-arch maybe you should include some checksums (md5?) in the repo so the data can be verified before flashing?

**Comment by stefanvanherwijnen** [comment]
My Neo3 is still running fine after 4 days with the new bootloader so the kernel panics seem to have been solved.

The checksums are a good idea. Or atleast add a mention that one should clone the repo instead of downloading the files manually. In theory it should make no difference, but as you said there is no way to check what went wrong.

**Comment by Pauligrinder** [comment]
Blah, gotta buy me a proper TTY-adapter. I tried connecting the NanoPi to a Rasperry Pi, but the only output I got were some white boxes, and even those were very few and far between, and seemed to have to do with some voltage fluctuations. 
I also tried the AVR-JTAG adapter I have into my MacBook, but that didn't print anything (tried cu, screen and nanocom)... I'm assuming I connected it right though, because the activity LED on it went absolutely nuts after plugging power into the NanoPi. I think at least the JTAG-adapter doesn't support high enough baud rates (it's pretty old), and the Raspberry Pi, I don't know, maybe the serial port wasn't enabled properly or something. All the instructions said to enable it using rapsiconfig, which I don't have since I'm running Arch on it.

Anyway, I gave up once again, as it just won't boot. Going back to my friendlycore-bootloader again and will get back to it once I get a proper serial adapter...

By the way guys, I know at least norman-arch is running the 2gb model of the Neo3 - what about you @stefanvanherwijnen? Are you running 1 or 2 gb? Mine is the 1gb model. It's really the only difference I can think of... 
Oh and this time I cloned the repo, so that was not the problem.

**Comment by Pauligrinder** [comment]
Well, guess what. Turns out the issue was with my sdcard reader/writer! I was using the integrated one in my old laptop, and as I was once again messing around with stuff, I happened to notice some input/output errors with it. They weren't even printed in DD output, but in dmesg! ... So I grabbed a 4G dongle that also has a microSD slot and put it in there, did all the same stuff, and bam. It works :)
